concordium-smart-contract-testing/src/invocation/impls.rs

```diff
@@ -0,0 +1,1468 @@
+                if *addr == invoked_contract {
+                    invoked_contract_has_state_changes = true;
+                }
+                modified_state.freeze(&mut loader, &mut collector);
+            }
+        }
+
+        // One energy per extra byte in the state.
+        let energy_for_state_increase = Energy::from(collector.collect());
+
+        if remaining_energy
+            .checked_sub(energy_for_state_increase)
+            .is_none()
+        {
+            return Err(OutOfEnergy);
+        }
+        Ok((
+            energy_for_state_increase,
+            invoked_contract_has_state_changes,
+        ))
+    }
+}
+
+impl Default for ChangeSet {
+    fn default() -> Self { Self::new() }
+}
+
+impl AmountDelta {
+    /// Create a new [`Self`], with the value `+0`.
+    fn new() -> Self { Self::Positive(Amount::zero()) }
+
+    /// Subtract an [`Amount`] from [`Self`].
+    fn subtract_amount(self, amount: Amount) -> Self {
+        match self {
+            Self::Positive(current) => {
+                if current >= amount {
+                    Self::Positive(current.subtract_micro_ccd(amount.micro_ccd))
+                } else {
+                    Self::Negative(amount.subtract_micro_ccd(current.micro_ccd))
+                }
+            }
+            Self::Negative(current) => Self::Negative(current.add_micro_ccd(amount.micro_ccd)),
+        }
+    }
+
+    /// Add an [`Amount`] from [`Self`].
+    fn add_amount(self, amount: Amount) -> Self {
+        match self {
+            Self::Positive(current) => Self::Positive(current.add_micro_ccd(amount.micro_ccd)),
+            Self::Negative(current) => {
+                if current >= amount {
+                    Self::Negative(current.subtract_micro_ccd(amount.micro_ccd))
+                } else {
+                    Self::Positive(amount.subtract_micro_ccd(current.micro_ccd))
+                }
+            }
+        }
+    }
+
+    /// Add two [`Self`] to create a new one.
+    fn add_delta(self, delta: AmountDelta) -> Self {
+        match delta {
+            AmountDelta::Positive(d) => self.add_amount(d),
+            AmountDelta::Negative(d) => self.subtract_amount(d),
+        }
+    }
+
+    /// Whether the [`Self`] is zero (either `+0` or `-0`).
+    fn is_zero(&self) -> bool {
+        match self {
+            AmountDelta::Positive(d) => d.micro_ccd == 0,
+            AmountDelta::Negative(d) => d.micro_ccd == 0,
+        }
+    }
+
+    /// Returns a new balance with the `AmountDelta` applied, or, an
+    /// error if `balance + self < 0`.
+    ///
+    /// Panics if an overflow occurs.
+    fn apply_to_balance(&self, balance: Amount) -> Result<Amount, UnderflowError> {
+        match self {
+            AmountDelta::Positive(d) => Ok(balance
+                .checked_add(*d)
+                .expect("Overflow occured when adding Amounts.")), // TODO: Should we return an
+            // error for this?
```

Comment:
I think in the node we assume this does not happen.

 Here in principle they can construct a state where this overflow can occur.

How annoying will it be to have a proper error here?